### PR TITLE
HubApi Popup Overlay

### PR DESCRIPTION
**Pull request by sisou**
Adds an overlay when the Hub is opened in a popup. Clicking on this overlay refocuses the popup.

Tested on Chrome (Desktop and Android), Firefox (Desktop), Safari (Mac). More tests are welcome!

I added some helper functions and variables so that the code can be better minified for the standalone builds (DOM variables & methods can't be minified, so abstracting them is the way to go).

Paypal Overlay for reference (click the "Try the buttons" buttons): https://developer.paypal.com/docs/checkout/

image

### CODE REVIEW

File: client/RequestBehavior.ts [hunk 1]
```diff
@@ -101,6 +108,81 @@ export class PopupRequestBehavior extends RequestBehavior<BehaviorType.POPUP> {
...
+        // can be animated, but also because it actually feels better when there is a short
+        // delay between the opening popup and the background fading.
+        setTimeout(() => overlay.style.opacity = '1', 100);
+
+        return appendChild(document.body, overlay) as HTMLDivElement;
+    }
+
+    private removeOverlay(overlay: HTMLDivElement): void {
+        overlay.style.opacity = '0';
+        setTimeout(() => document.body.removeChild(overlay), 400);
```
Comment (nibhar) [inline client/RequestBehavior.ts]: Easing function on opacity runs for `.6`, why `400` here?

**Comment by sisou** [comment]
Because it will be used on all websites that use the HubApi, not only Nimiq apps, so we have no control over the `rem` size.

File: client/RequestBehavior.ts (hunk 1 above)
Comment (sisou) [reply client/RequestBehavior.ts]: So the website becomes interactive earlier, when the overlay is barely visible anymore anyway.

File: client/RequestBehavior.ts [hunk 2]
```diff
@@ -101,6 +108,81 @@ export class PopupRequestBehavior extends RequestBehavior<BehaviorType.POPUP> {
...
+        overlayStyle.transition = 'opacity 0.6s ease';
+        overlay.addEventListener('click', () => popup.focus());
+
+        // Top flex spacer
+        appendChild(overlay, createElement('div'));
+
+        // Explainer text
+        const text = createElement('div');
+        const textStyle = text.style;
+        appendChild(text, createTextNode('A popup is active,'));
```
Comment (danimoh) [inline client/RequestBehavior.ts]: Sounds weird to me.
How about 'A popup has been opened'?

File: client/RequestBehavior.ts [hunk 3]
```diff
@@ -101,6 +108,81 @@ export class PopupRequestBehavior extends RequestBehavior<BehaviorType.POPUP> {
...
+        appendChild(text, createElement('br'));
+        appendChild(text, createTextNode('click anywhere to refocus.'));
```
Comment (danimoh) [inline client/RequestBehavior.ts]: 'click anywhere to bring it back to the front.'?

File: client/RequestBehavior.ts [hunk 4]
```diff
@@ -101,6 +108,81 @@ export class PopupRequestBehavior extends RequestBehavior<BehaviorType.POPUP> {
...
+        textStyle.fontSize = '24px';
+        textStyle.fontWeight = '600';
+        textStyle.lineHeight = '40px';
```
Comment (danimoh) [inline client/RequestBehavior.ts]: If Muli is available on the page, I think we should use it. I.e.:
`textStyle.fontFamily = 'Muli', fallbacks...`

**Pull request merged**